include/openssl/core_numbers.h

```diff
@@ -333,6 +334,54 @@ OSSL_CORE_MAKE_FUNC(int, OP_kdf_get_ctx_params,
 OSSL_CORE_MAKE_FUNC(int, OP_kdf_set_ctx_params,
                     (void *kctx, const OSSL_PARAM params[]))
 
+/* RAND */
+
+# define OSSL_FUNC_RAND_NEWCTX                        1
+# define OSSL_FUNC_RAND_FREECTX                       2
+# define OSSL_FUNC_RAND_INSTANTIATE                   3
+# define OSSL_FUNC_RAND_UNINSTANTIATE                 4
+# define OSSL_FUNC_RAND_GENERATE                      5
+# define OSSL_FUNC_RAND_GET_NONCE                     6
+# define OSSL_FUNC_RAND_RESEED                        7
+# define OSSL_FUNC_RAND_GETTABLE_PARAMS               8
+# define OSSL_FUNC_RAND_GETTABLE_CTX_PARAMS           9
+# define OSSL_FUNC_RAND_SETTABLE_CTX_PARAMS          10
+# define OSSL_FUNC_RAND_GET_PARAMS                   11
+# define OSSL_FUNC_RAND_GET_CTX_PARAMS               12
+# define OSSL_FUNC_RAND_SET_CTX_PARAMS               13
+# define OSSL_FUNC_RAND_SET_CALLBACKS                14
+
+OSSL_CORE_MAKE_FUNC(int, OP_rand_get_bytes,
+                    (void *vctx, unsigned char *out, size_t outlen,
+                     const unsigned char *addin, size_t addin_len))
+OSSL_CORE_MAKE_FUNC(int, OP_rand_get_nonce,
+                    (void *vctx, unsigned char *out, size_t outlen))
+OSSL_CORE_MAKE_FUNC(int, OP_rand_reseed,
+                    (void *vctx, const unsigned char *ent, size_t ent_len,
+                     const unsigned char *addin, size_t addin_len))
+OSSL_CORE_MAKE_FUNC(void *, OP_rand_newctx,
+                    (void *provctx, int secure, int df,
+                     OSSL_OP_rand_get_bytes_fn *parent_get_bytes,
+                     OSSL_OP_rand_reseed_fn *parent_reseed, void *parent))
```

Comment:
I'd like to understand the line of thinking a bit more.  I assume this is to allow a connection between "this" EVP_RAND implementation and the parent, and that the two functions passed are the parent functions needed for the two instances to talk with each other, and further that this allows two EVP_RAND *from different providers* to do so, is that correct?

I'm a bit nervous about making a hard dependence on the function signatures, and also in assuming that we know what each implementation needs.  Since those functions must come from an EVP_RAND fetched in libcrypto, we could as well pass the whole OSSL_DISPATCH and let the implementation pilfer whatever it needs, rather than just a few functions.  Of course, that means that EVP_RAND will have to store a pointer to said OSSL_DISPATCH, but that shouldn't be too hard, right?

Reply:
The basic problem is that we need to connect DRBGs into a chain.  This is something that I believe is better done by libcrypto than the provider.  By having it in libcrypto, we retain the flexibility to alter our DRBG architecture in the future.  Unfortunately, FIPS requires that any data passed between chained DRBGs is either done using a secure channel or the data stays within the FIPS boundary.  A secure channel requires encryption and checksumming -- too hard.  This means that the data needs to stay inside.  Therefore, each DRBG needs to be able to directly call its parent for entropy.

Yes, it would allow EVP_RAND from different providers to be chained.  There is a  use case for this: it permits a validated HRNG to seed our FIPS DRBGs.  Agreed that this isn't essential for our validation but it is something that is desirable.

Passing the entire OSSL_DISPATCH is an option I hadn't considered.  It makes sense.
Using OSSL_CALLBACK for this seemed really awkward.


It would also be possible to reduce the number of functions required to one by added a _prediction resistance_ argument to the _get bytes_ call.  I might still do this but it requires more code changes.

Reply:
Locking is also required.  Either in the provider or as callbacks to EVP.